base/abstractarray.jl

```diff
@@ -676,7 +696,7 @@ function hcat{T}(A::AbstractVecOrMat{T}...)
     for j = 1:nargs
         Aj = A[j]
         if size(Aj, 1) != nrows
-            throw(ArgumentError("number of rows must match"))
+            throw(ArgumentError("number of rows of A[$j] ($(size(Aj,1))) must match $nrows"))
```

Comment:
Better to avoid the `A` in the error message since it doesn't mean anything to the user of the function. Printing the object is probably too much, just print out the index of the argument is probably ok.